### PR TITLE
Validate annotation processors in maven-compiler-plugin

### PR DESCRIPTION
**Pull request by cobratbq**
This PR adds support for validating the annotation processors that might be listed in the maven-compiler-plugin configuration. These annotation processors are basically dependencies, but referenced through the configuration, instead of dependency listing.

- adds configuration parameter `verifyAtypical`, which currently only checks `maven-compiler-plugin` configuration for annotation processors, but can in the future contain other "exotic" locations.
- Fixes issue with dependencies filter being used for build plug-ins: filtering by scope is dangerous for plugins as these are all `runtime` artifacts.
- Created `Configuration` container such that all behavioral properties of the `ArtifactResolver` are combined in a single parameter.
- includes unit tests and IT tests

NOTE: As with build plugins, we currently only validate the referenced artifacts themselves. Their dependencies are ignored. (Though once we solve resolution, we can apply it in both cases.)

### CODE REVIEW

File: src/main/java/org/simplify4u/plugins/ArtifactResolver.java [hunk 1]
```diff
@@ -112,31 +114,46 @@ private ArtifactRepositoryPolicy policyIgnoreCheckSum(ArtifactRepositoryPolicy p
...
      *
      * @return Returns set of all artifacts whose signature needs to be verified.
      */
     Set<Artifact> resolveProjectArtifacts(MavenProject project, SkipFilter filter, boolean verifyPomFiles,
-                                          boolean verifyPlugins) throws MojoExecutionException {
+                                          boolean verifyPlugins, boolean verifyAtypical) throws MojoExecutionException {
         final LinkedHashSet<Artifact> allArtifacts = new LinkedHashSet<>(
                 resolveArtifacts(project.getArtifacts(), filter, verifyPomFiles));
         if (verifyPlugins) {
+            // FIXME: applying dependency filters to build plug-ins may result in all build plug-ins being dropped because these are listed as 'runtime'-scoped.
```
Comment (cobratbq) [inline src/main/java/org/simplify4u/plugins/ArtifactResolver.java]: @slawekjaranowski can you double-check if you really wanted the `filter` parameter to be changed in your earlier commit?

I think that this has unintended side-effects. Most significantly: if scope is reduced to `compile` or such, no plugins will be verified due to plugins having `runtime` scope by default.

Comment (slawekjaranowski) [reply src/main/java/org/simplify4u/plugins/ArtifactResolver.java]: my intention was to filter snapshot versions, maybe only snapshot should be filtered

File: src/test/java/org/simplify4u/plugins/MavenCompilerUtilsTest.java [hunk 2]
```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright 2017 Danny van Heumen
```
Comment (slawekjaranowski) [inline src/test/java/org/simplify4u/plugins/MavenCompilerUtilsTest.java]: put current year

File: src/main/java/org/simplify4u/plugins/MavenCompilerUtils.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2017 Danny van Heumen
```
Comment (slawekjaranowski) [inline src/main/java/org/simplify4u/plugins/MavenCompilerUtils.java]: put current year

File: src/main/java/org/simplify4u/plugins/ArtifactResolver.java (hunk 1 above)
Comment (cobratbq) [reply src/main/java/org/simplify4u/plugins/ArtifactResolver.java]: Thanks for confirming. I'll fix it in the PR. I'll provide a separate skipfilter for plugins which contains just the snapshot-filter or nothing.

File: src/test/java/org/simplify4u/plugins/MavenCompilerUtilsTest.java (hunk 2 above)
Comment (cobratbq) [reply src/test/java/org/simplify4u/plugins/MavenCompilerUtilsTest.java]: I feel like an idiot for making these kinds of trivial mistakes :-/ Thanks for catching this!

File: src/main/java/org/simplify4u/plugins/MavenCompilerUtils.java (hunk 3 above)
Comment (cobratbq) [reply src/main/java/org/simplify4u/plugins/MavenCompilerUtils.java]: Thanks for catching this!

**Comment by cobratbq** [comment]
I ran the merge with master. It merges without issues. I did not push the merge-commit, in case you want to keep the history graph clean (relatively speaking).

**Pull request merged**